[LLVM][NVPTX] Add NVPTX codegen support for clusterlaunchcontrol instruction (#134568)
llvm/llvm-project: llvm/docs/NVPTXUsage.rst

For more information, refer `PTX ISA <https://docs.nvidia.com/cuda/parallel-thread-execution/#data-movement-and-conversion-instructions-st-bulk>`__.

clusterlaunchcontrol Intrinsics
-------------------------------

'``llvm.nvvm.clusterlaunchcontrol.try_cancel*``' Intrinsics
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Syntax:
"""""""

.. code-block:: llvm

  declare void @llvm.nvvm.clusterlaunchcontrol.try_cancel.async(ptr %addr, ptr %mbar)
  declare void @llvm.nvvm.clusterlaunchcontrol.try_cancel.async.shared(ptr addrspace(3) %addr, ptr addrspace(3) %mbar)
  declare void @llvm.nvvm.clusterlaunchcontrol.try_cancel.async.multicast(ptr %addr, ptr %mbar)

>   declare void @llvm.nvvm.clusterlaunchcontrol.try_cancel.async.multicast.shared(ptr addrspace(3) %addr, ptr addrspace(3) %mbar)

Author: That's right. We have used `shared` in the past to refer to `shared::cta` this keeps the name concise. If the instruction extends support for shared::cluster, we can add shared_cluster variants which would disambiguate. Also, NVPTX now support addrspace for shared::cluster, so the operand would disambiguate as well

Overview:
"""""""""

The ``clusterlaunchcontrol.try_cancel`` intrinsics requests atomically cancelling

> the launch of a cluster that has not started running yet. It asynchronously writes

Reviewer: ```suggestion
the launch of a cluster that has not started running yet. It asynchronously non-atomically writes
```

Author: Updated the doc with the suggested change

a 16-byte opaque response to shared memory, pointed to by ``addr`` indicating whether the

> operation succeeded or failed. ``addr`` and ``mbar`` must be in ``shared::cta``

Reviewer: ```suggestion
operation succeeded or failed. ``addr`` and ``mbar`` must be in ``shared::cta``,
```

Author: Updated the doc with the suggested change

> otherwise the result is undefined. The completion of the asynchronous operation

Reviewer: We don't guarantee that a result will be written at all:

```suggestion
otherwise the behavior is undefined. The completion of the asynchronous operation
```

Author: Updated the doc with the suggested change

is tracked using the mbarrier completion mechanism at ``.cluster`` scope referenced
by the shared memory pointer, ``mbar``. On success, the opaque response contains
the CTA id of the first CTA of the canceled cluster; no other successful response
from other ``clusterlaunchcontrol.try_cancel`` operations from the same grid will
contain that id.

> The ``multicast`` variant specifies that the response is asynchronously written to

Reviewer: ```suggestion
The ``multicast`` variant specifies that the response is asynchronously non-atomically written to
```

the corresponding shared memory location of each CTA in the requesting cluster.
The completion of the write of each local response is tracked by independent
mbarriers at the corresponding shared memory location of each CTA in the
cluster.

For more information, refer `PTX ISA <https://docs.nvidia.com/cuda/parallel-thread-execution/?a#parallel-synchronization-and-communication-instructions-clusterlaunchcontrol-try-cancel>`__.

'``llvm.nvvm.clusterlaunchcontrol.query_cancel.is_canceled``' Intrinsic
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Syntax:
"""""""

.. code-block:: llvm

  declare i1 @llvm.nvvm.clusterlaunchcontrol.query_cancel.is_canceled(i128 %try_cancel_response)

Overview:
"""""""""

> The ``llvm.nvvm.clusterlaunchcontrol.query_cancel.is_canceled`` intrinsic can be
> used to decode the opaque response written by the

Reviewer: ```suggestion
The ``llvm.nvvm.clusterlaunchcontrol.query_cancel.is_canceled`` intrinsic
decodes the opaque response written by the
```

Author: Updated the doc with the suggested change

``llvm.nvvm.clusterlaunchcontrol.try_cancel`` operation.

> The intrinsic returns false if the request failed. If the request succeeded,

Reviewer: ```suggestion
The intrinsic returns ``0`` (false) if the request failed. If the request succeeded,
```

> it returns true. A true result indicates that:

Reviewer: ```suggestion
it returns ``1`` (true). A true result indicates that:
```

- the thread block cluster whose first CTA id matches that of the response

>   handle will not run

Author: Updated the doc with the suggested change

> - no other successful response of another ``try_cancel`` request will contain

Reviewer: ```suggestion
- no other successful response of another ``try_cancel`` request in the grid will contain
```

Author: Updated the doc with the suggested change

  the first CTA id of that cluster

For more information, refer `PTX ISA <https://docs.nvidia.com/cuda/parallel-thread-execution/?a#parallel-synchronization-and-communication-instructions-clusterlaunchcontrol-query-cancel>`__.

'``llvm.nvvm.clusterlaunchcontrol.query_cancel.get_first_ctaid.*``' Intrinsics
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Syntax:
"""""""

.. code-block:: llvm

  declare <4 x i32> @llvm.nvvm.clusterlaunchcontrol.query_cancel.get_first_ctaid(i128 %try_cancel_response)
  declare i32 @llvm.nvvm.clusterlaunchcontrol.query_cancel.get_first_ctaid.x(i128 %try_cancel_response)
  declare i32 @llvm.nvvm.clusterlaunchcontrol.query_cancel.get_first_ctaid.y(i128 %try_cancel_response)
  declare i32 @llvm.nvvm.clusterlaunchcontrol.query_cancel.get_first_ctaid.z(i128 %try_cancel_response)

Overview:
"""""""""

The ``clusterlaunchcontrol.query_cancel.get_first_ctaid`` intrinsic can be
used to decode the opaque response written by the
``llvm.nvvm.clusterlaunchcontrol.try_cancel`` operation.

> If the request succeeded,
> ``llvm.nvvm.clusterlaunchcontrol.query_cancel.get_first_ctaid.{x,y,z}`` returns
> the coordinate of the first CTA in the canceled cluster, either x, y, or z.
> ``llvm.nvvm.clusterlaunchcontrol.query_cancel.get_first_ctaid`` returns a vector
> where the first three components are the x, y, z coordinates of the first CTA.
> The contents of the fourth element are unspecified

Reviewer: ```suggestion
If the request succeeded,

    - ``llvm.nvvm.clusterlaunchcontrol.query_cancel.get_first_ctaid.{x,y,z}`` returns
      the coordinate of the first CTA in the canceled cluster, either x, y, or z.
    - ``llvm.nvvm.clusterlaunchcontrol.query_cancel.get_first_ctaid`` returns a vector
      where the first three components are the x, y, z coordinates of the first CTA.

The contents of the fourth element are unspecified.
```

Author: Updated the doc with the suggested change
